tpm2/structures.go

```diff
@@ -371,3 +371,158 @@ type tpmtSigScheme struct {
+	}
+	n, err := decodeName(buf)
+	if err != nil {
+		return nil, fmt.Errorf("decoding QualifiedSigner: %v", err)
+	}
+	ad.QualifiedSigner = *n
+	if err := tpmutil.UnpackBuf(buf,
+		&ad.ExtraData,
+		&ad.ClockInfo,
+		&ad.FirmwareVersion,
+	); err != nil {
+		return nil, fmt.Errorf("decoding ExtraData/ClockInfo/FirmwareVersion: %v", err)
+	}
+
+	// Add support for more types as needed.
+	if ad.Type != tagAttestCertify {
+		return nil, fmt.Errorf("only Certify attestation structure is supported, got type 0x%x", ad.Type)
+	}
+	ad.AttestedCertifyInfo, err = decodeCertifyInfo(buf)
+	if err != nil {
+		return nil, fmt.Errorf("decoding AttestedCertifyInfo: %v", err)
+	}
+	return &ad, nil
+}
+
+// CertifyInfo contains Certify-specific data for TPMS_ATTEST.
+type CertifyInfo struct {
+	Name          Name
+	QualifiedName Name
+}
+
+func decodeCertifyInfo(in *bytes.Buffer) (*CertifyInfo, error) {
+	var ci CertifyInfo
+
+	n, err := decodeName(in)
+	if err != nil {
+		return nil, fmt.Errorf("decoding Name: %v", err)
+	}
+	ci.Name = *n
+
+	n, err = decodeName(in)
+	if err != nil {
+		return nil, fmt.Errorf("decoding QualifiedName: %v", err)
+	}
+	ci.QualifiedName = *n
+
+	return &ci, nil
+}
+
+// Name contains a Name for TPM entities. This is either a handle or a digest.
+// Only one of Handle/Digest will be set.
+type Name struct {
+	Handle *tpmutil.Handle
+	Digest *HashValue
+}
+
+func decodeName(in *bytes.Buffer) (*Name, error) {
+	var nameBuf []byte
+	if err := tpmutil.UnpackBuf(in, &nameBuf); err != nil {
+		return nil, err
+	}
+
+	name := new(Name)
+	switch len(nameBuf) {
+	case 0:
+		// no name is present
+	case 4:
+		name.Handle = new(tpmutil.Handle)
+		if err := tpmutil.UnpackBuf(bytes.NewBuffer(nameBuf), name.Handle); err != nil {
+			return nil, fmt.Errorf("decoding Handle: %v", err)
+		}
+	default:
+		var err error
+		name.Digest, err = decodeHashValue(bytes.NewBuffer(nameBuf))
+		if err != nil {
+			return nil, fmt.Errorf("decoding Digest: %v", err)
+		}
+	}
+	return name, nil
+}
+
+// MatchesPublic compares Digest in Name against given Public structure. Note:
+// this only works for regular Names, not Qualified Names.
+func (n Name) MatchesPublic(p Public) (bool, error) {
+	buf, err := p.encode()
+	if err != nil {
+		return false, err
+	}
+	if n.Digest == nil {
+		return false, errors.New("Name doesn't have a Digest, can't compare to Public")
+	}
+	hfn, ok := toGoHashes[n.Digest.Alg]
+	if !ok {
+		return false, fmt.Errorf("Name hash algorithm 0x%x not supported", n.Digest.Alg)
+	}
+
+	h := hfn()
+	h.Write(buf)
+	digest := h.Sum(nil)
+
+	return bytes.Equal(digest, n.Digest.Value), nil
+}
+
+// HashValue is an algorithm-specific hash value.
+type HashValue struct {
+	Alg   Algorithm
+	Value []byte
+}
+
+func decodeHashValue(in *bytes.Buffer) (*HashValue, error) {
+	var hv HashValue
+	if err := tpmutil.UnpackBuf(in, &hv.Alg); err != nil {
+		return nil, fmt.Errorf("decoding Alg: %v", err)
+	}
+	h, ok := toGoHashes[hv.Alg]
```

Comment:
For consistency with the previous code, this should be hfn, since h was used there as the result of calling the function returned from toGoHashes

Reply:
Done